### PR TITLE
#Week20 Fullstack project (Jessika & Laima)

### PR DESCRIPTION
**Pull request by leelalaim**
Link to my deployed project:
* Backend: https://jokeio.herokuapp.com/
* Frontend: https://jokeio.netlify.app/

My reflections on how this week's project turned out:
It was an awesome week :)

### CODE REVIEW

**Review by sssylvasss — APPROVED** [review]
Wow-what a long code;) I love your page! 

I did have some problem logging in, my name was too long, and my username too short. Once I was in I read some jokes and logged out again. Then I remember that I wanted to leave a comment for the next user but I could not log in again? I probably messed something up but I tried again to get a new account but then I needed one other email address and there was no way to resend the password. What I want to say is that all your error messages work! I got them all!!:) 

Well done!!!
/y&m

File: backend/server.js [hunk 1]
```diff
@@ -1,30 +1,182 @@
...
+  password: {
+    type: String,
+    required: [true, "You need to choose a password"],
+    minlength: [8, "Your password needs to contain min 8 characters"],
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString("hex"),
+  },
+});
```
Comment (KarolCislowski) [inline backend/server.js:52]: Awesome model 😻

File: backend/server.js [hunk 2]
```diff
@@ -1,30 +1,182 @@
...
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(express.json());
+
+// Routes
```
Comment (KarolCislowski) [inline backend/server.js:88]: nice set of routes 👍🏻

File: frontend/src/App.js [hunk 3]
```diff
@@ -1,9 +1,63 @@
...
+
+const store = createStore(
+  reducer,
+  persistedState,
+  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
+);
+
+store.subscribe(() => {
+  localStorage.setItem("reduxState", JSON.stringify(store.getState()));
+});
```
Comment (KarolCislowski) [inline frontend/src/App.js:46]: 😻

File: frontend/src/pages/Joke.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  useEffect(() => {
+    if (!accessToken) {
+      history.push("/signin");
+    }
+  }, [accessToken, history]);
+
+  useEffect(() => {
+    fetchJoke();
+  }, []);
```
Comment (KarolCislowski) [inline frontend/src/pages/Joke.js:35]:
```jsx 
useEffect(() => {
    if (!accessToken) {
      history.push("/signin");
    } else {
      fetchJoke();
    }
  }, [accessToken, history]);
```
Why to effects when you can have one?

File: frontend/src/pages/Main.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  MessageContainer,
+  ReceivedMessageContainer,
+  IconContainer,
+  Heart,
+  MessageTitle,
+  Message,
+  Form,
+  InputMessage,
+  MessageButton,
+  SentMessageText,
```
Comment (KarolCislowski) [inline frontend/src/pages/Main.js:22]: I love how you approach styling in your project 😻

File: frontend/src/pages/Main.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  }, [accessToken, history]);
+
+  useEffect(() => {
+    fetch(API_URL("usermessage"))
+      .then((res) => res.json())
+      .then((json) => {
+        setMessages(json.userMessage[0].message);
+      })
+      .catch();
+  }, []);
```
Comment (KarolCislowski) [inline frontend/src/pages/Main.js:48]: again you can create one effect from it.

File: frontend/src/pages/SignUp.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      setErrorMessage(error.message);
+    } else if (errorType.username) {
+      setErrorMessage(errorType.username.message);
+    } else if (errorType.name) {
+      setErrorMessage(errorType.name.message);
+    } else if (errorType.email) {
+      setErrorMessage(errorType.email.message);
+    } else {
+      setErrorMessage(errorType);
+    }
```
Comment (KarolCislowski) [inline frontend/src/pages/SignUp.js:94]: nice error handling

**Review by KarolCislowski — APPROVED** [review]
Beautiful and fully functional page! you can be proud of yourself!

**Pull request closed without merging**